contracts/utils/Roles.sol

```diff
@@ -0,0 +1,281 @@
+     */
+    function _setProposedRoles(
+        address fromAccount,
+        address toAccount,
+        uint256 proposedRoles_
+    ) private {
+        bytes32 key = _proposalKey(fromAccount, toAccount);
+        assembly {
+            sstore(key, proposedRoles_)
+        }
+    }
+
+    /**
+     * @notice Internal function to add roles to an account.
+     * @dev emits a RolesAdded event.
+     * @dev Called in the constructor to set the initial roles.
+     * @param account The address to add roles to
+     * @param roles The roles to add
+     */
+    function _addRoles(address account, uint8[] memory roles) internal {
+        uint256 accountRoles = _getRoles(account);
+        uint256 length = roles.length;
+
+        for (uint256 i = 0; i < length; ++i) {
+            accountRoles |= (1 << roles[i]);
+        }
+
+        _setRoles(account, accountRoles);
+
+        emit RolesAdded(account, roles);
+    }
+
+    /**
+     * @notice Internal function to remove roles from an account.
+     * @dev emits a RolesRemoved event.
+     * @param account The address to remove roles from
+     * @param roles The roles to remove
+     */
+    function _removeRoles(address account, uint8[] memory roles) internal {
+        uint256 accountRoles = _getRoles(account);
+        uint256 length = roles.length;
+
+        for (uint256 i = 0; i < length; ++i) {
+            accountRoles &= ~(1 << roles[i]);
+        }
+
+        _setRoles(account, accountRoles);
+
+        emit RolesRemoved(account, roles);
+    }
+
+    /**
+     * @notice Internal function to check if an account has a role.
+     * @param accountRoles The roles of the account in uint256 format
+     * @param role The role to check
+     * @return True if the account has the role, false otherwise
+     */
+    function _hasRole(uint256 accountRoles, uint8 role) internal pure returns (bool) {
+        return accountRoles & (1 << role) != 0;
+    }
+
+    /**
+     * @notice Internal function to check if an account has all the roles.
+     * @param accountRoles The roles of the account in uint256 format
+     * @param roles The roles to check
+     * @return True if the account has all the roles, false otherwise
+     */
+    function _hasAllTheRoles(uint256 accountRoles, uint8[] memory roles) internal pure returns (bool) {
+        uint256 length = roles.length;
+
+        for (uint256 i = 0; i < length; ++i) {
+            if (accountRoles & (1 << roles[i]) == 0) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * @notice Internal function to check if an account has any of the roles.
+     * @param accountRoles The roles of the account in uint256 format
+     * @param roles The roles to check
+     * @return True if the account has any of the roles, false otherwise
+     */
+    function _hasAnyOfRoles(uint256 accountRoles, uint8[] memory roles) internal pure returns (bool) {
```

Comment:
don't think we need to add this due to bytecode size increase

Reply:
It's not included in the bytecode if inheriting contract is not using it.